src/query/api/v1/handler/prom/read.go

```diff
@@ -40,94 +41,136 @@ import (
 	"go.uber.org/zap"
 )
 
+// NewQueryFn creates a new promql Query.
+type NewQueryFn func(
+	engine *promql.Engine,
+	queryable promstorage.Queryable,
+	params models.RequestParams,
+) (promql.Query, error)
+
+var (
+	newRangeQueryFn = func(
+		engine *promql.Engine,
+		queryable promstorage.Queryable,
+		params models.RequestParams,
+	) (promql.Query, error) {
+		return engine.NewRangeQuery(
+			queryable,
+			params.Query,
+			params.Start,
+			params.End,
+			params.Step)
+	}
+
+	newInstantQueryFn = func(
+		engine *promql.Engine,
+		queryable promstorage.Queryable,
+		params models.RequestParams,
+	) (promql.Query, error) {
+		return engine.NewInstantQuery(
+			queryable,
+			params.Query,
+			params.Now)
+	}
+)
+
 type readHandler struct {
-	engine    *promql.Engine
-	queryable promstorage.Queryable
-	hOpts     options.HandlerOptions
-	scope     tally.Scope
-	logger    *zap.Logger
+	engine     *promql.Engine
+	queryable  promstorage.Queryable
+	hOpts      options.HandlerOptions
+	scope      tally.Scope
+	logger     *zap.Logger
+	opts       opts
 }
 
 func newReadHandler(
-	opts Options,
 	hOpts options.HandlerOptions,
 	queryable promstorage.Queryable,
-) http.Handler {
+	options ...Option,
+) (http.Handler, error) {
 	scope := hOpts.InstrumentOpts().MetricsScope().Tagged(
 		map[string]string{"handler": "prometheus-read"},
 	)
-	return &readHandler{
-		engine:    opts.PromQLEngine,
-		queryable: queryable,
-		hOpts:     hOpts,
-		scope:     scope,
-		logger:    hOpts.InstrumentOpts().Logger(),
+	opts := newDefaultOptions(hOpts)
+	for _, optionFn := range options {
+		if err := optionFn(&opts); err != nil {
+			return nil, err
+		}
 	}
+
+	return &readHandler{
+		engine:     opts.promQLEngine,
+		queryable:  queryable,
```

Comment:
Unless I'm missing something, both `engine` and `queryable`  seems to be used to construct an instance of `promql.Query`. If we're switching to injecting a function/closure for it, we might as well get rid of these two fields by using them in the `newQueryFn` closure. E.g.:

```go
type newQueryFn func(models.RequestParams) (promql.Query, error)

type readHandler struct {
    newQueryFn newQueryFn
    hOpts      options.HandlerOptions
    scope      tally.Scope
    logger     *zap.Logger
    /* ... */
}

func newRangeQueryFn(engine *promql.Engine, queryable  promstorage.Queryable) newQueryFn {
    return func(params models.RequestParams) (promql.Query, error) {
        return engine.NewRangeQuery(queryable, /* params.* */)
    }
}
```